<a href="https://www.kaggle.com/parisrohan/rohan-house-price-prediction?scriptVersionId=89922433" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/house-prices-advanced-regression-techniques/sample_submission.csv
/kaggle/input/house-prices-advanced-regression-techniques/data_description.txt
/kaggle/input/house-prices-advanced-regression-techniques/train.csv
/kaggle/input/house-prices-advanced-regression-techniques/test.csv


In [2]:
#load required libraries
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

pd.pandas.set_option('display.max_columns',None)

# Load Data

In [3]:
df_train=pd.read_csv('/kaggle/input/house-prices-advanced-regression-techniques/train.csv')
df_train.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
1         Lvl    AllPub       FR2       Gtl      Veenker      Feedr   
2         Lvl    AllPub    Inside       Gtl      CollgCr       Norm   
3         Lvl    AllPub    Corner       Gtl      Crawfor       Norm   
4         Lvl    AllPub       FR2       Gtl      NoRidge       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     2Story            7            5       2003   
1       Norm     1Fam     1Story            6            8       1976   
2       Norm     1Fam     2Story            7            5       2001   
3       Norm     1Fam     2Story            7            5       1915   
4       Norm     1Fam     2Story            8            5       2000   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          2003     Gable  CompShg     VinylSd     VinylSd    BrkFace   
1          1976     Gable  CompShg     MetalSd     MetalSd       None   
2          2002     Gable  CompShg     VinylSd     VinylSd    BrkFace   
3          1970     Gable  CompShg     Wd Sdng     Wd Shng       None   
4          2000     Gable  CompShg     VinylSd     VinylSd    BrkFace   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       196.0        Gd        TA      PConc       Gd       TA           No   
1         0.0        TA        TA     CBlock       Gd       TA           Gd   
2       162.0        Gd        TA      PConc       Gd       TA           Mn   
3         0.0        TA        TA     BrkTil       TA       Gd           No   
4       350.0        Gd        TA      PConc       Gd       TA           Av   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          GLQ         706          Unf           0        150          856   
1          ALQ         978          Unf           0        284         1262   
2          GLQ         486          Unf           0        434          920   
3          ALQ         216          Unf           0        540          756   
4          GLQ         655          Unf           0        490         1145   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        Ex          Y      SBrkr       856       854             0   
1    GasA        Ex          Y      SBrkr      1262         0             0   
2    GasA        Ex          Y      SBrkr       920       866             0   
3    GasA        Gd          Y      SBrkr       961       756             0   
4    GasA        Ex          Y      SBrkr      1145      1053             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0       1710             1             0         2         1             3   
1       1262             0             1         2         0             3   
2       1786             1             0         2         1             3   
3       1717             1             0         1         0             3   
4       2198             1             0         2         1             4   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          Gd             8        Typ           0         NaN   
1             1          TA             6        Typ           1          TA   
2             1          Gd             6        Typ           1          TA   
3             1          Gd            

In [4]:
df_test=pd.read_csv('/kaggle/input/house-prices-advanced-regression-techniques/test.csv')
df_test.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl        NAmes      Feedr   
1         Lvl    AllPub    Corner       Gtl        NAmes       Norm   
2         Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
3         Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
4         HLS    AllPub    Inside       Gtl      StoneBr       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     1Story            5            6       1961   
1       Norm     1Fam     1Story            6            6       1958   
2       Norm     1Fam     2Story            5            5       1997   
3       Norm     1Fam     2Story            6            6       1998   
4       Norm   TwnhsE     1Story            8            5       1992   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          1961     Gable  CompShg     VinylSd     VinylSd       None   
1          1958       Hip  CompShg     Wd Sdng     Wd Sdng    BrkFace   
2          1998     Gable  CompShg     VinylSd     VinylSd       None   
3          1998     Gable  CompShg     VinylSd     VinylSd    BrkFace   
4          1992     Gable  CompShg     HdBoard     HdBoard       None   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0         0.0        TA        TA     CBlock       TA       TA           No   
1       108.0        TA        TA     CBlock       TA       TA           No   
2         0.0        TA        TA      PConc       Gd       TA           No   
3        20.0        TA        TA      PConc       TA       TA           No   
4         0.0        Gd        TA      PConc       Gd       TA           No   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          Rec       468.0          LwQ       144.0      270.0        882.0   
1          ALQ       923.0          Unf         0.0      406.0       1329.0   
2          GLQ       791.0          Unf         0.0      137.0        928.0   
3          GLQ       602.0          Unf         0.0      324.0        926.0   
4          ALQ       263.0          Unf         0.0     1017.0       1280.0   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        TA          Y      SBrkr       896         0             0   
1    GasA        TA          Y      SBrkr      1329         0             0   
2    GasA        Gd          Y      SBrkr       928       701             0   
3    GasA        Ex          Y      SBrkr       926       678             0   
4    GasA        Ex          Y      SBrkr      1280         0             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0        896           0.0           0.0         1         0             2   
1       1329           0.0           0.0         1         1             3   
2       1629           0.0           0.0         2         1             3   
3       1604           0.0           0.0         2         1             3   
4       1280           0.0           0.0         2         0             2   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          TA             5        Typ           0         NaN   
1             1          Gd             6        Typ           0         NaN   
2             1          TA             6        Typ           1          TA   
3             1          Gd  

In [5]:
df_train.shape, df_test.shape

((1460, 81), (1459, 80))

**Train** data has 1460 rows and 81 columns AND **Test** data has 1459 rows and 80 columns

# Data Cleaning

In [6]:
#Function to get count of missing values in each column
def get_cols_with_missing_values(DataFrame):
    missing_na_columns=(DataFrame.isnull().sum())
    return missing_na_columns[missing_na_columns > 0]

In [7]:
#Missing values in train data
feature_with_na = get_cols_with_missing_values(df_train)
print(feature_with_na)

LotFrontage      259
Alley           1369
MasVnrType         8
MasVnrArea         8
BsmtQual          37
BsmtCond          37
BsmtExposure      38
BsmtFinType1      37
BsmtFinType2      38
Electrical         1
FireplaceQu      690
GarageType        81
GarageYrBlt       81
GarageFinish      81
GarageQual        81
GarageCond        81
PoolQC          1453
Fence           1179
MiscFeature     1406
dtype: int64


The dependent feature 'SalePrice' has no missing values

In [8]:
#combine train and test data for data preprocessing
df_merge=pd.concat([df_test.assign(ind="test"), df_train.assign(ind="train")])
df_merge.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   

  LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0         Lvl    AllPub    Inside       Gtl        NAmes      Feedr   
1         Lvl    AllPub    Corner       Gtl        NAmes       Norm   
2         Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
3         Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
4         HLS    AllPub    Inside       Gtl      StoneBr       Norm   

  Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0       Norm     1Fam     1Story            5            6       1961   
1       Norm     1Fam     1Story            6            6       1958   
2       Norm     1Fam     2Story            5            5       1997   
3       Norm     1Fam     2Story            6            6       1998   
4       Norm   TwnhsE     1Story            8            5       1992   

   YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0          1961     Gable  CompShg     VinylSd     VinylSd       None   
1          1958       Hip  CompShg     Wd Sdng     Wd Sdng    BrkFace   
2          1998     Gable  CompShg     VinylSd     VinylSd       None   
3          1998     Gable  CompShg     VinylSd     VinylSd    BrkFace   
4          1992     Gable  CompShg     HdBoard     HdBoard       None   

   MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0         0.0        TA        TA     CBlock       TA       TA           No   
1       108.0        TA        TA     CBlock       TA       TA           No   
2         0.0        TA        TA      PConc       Gd       TA           No   
3        20.0        TA        TA      PConc       TA       TA           No   
4         0.0        Gd        TA      PConc       Gd       TA           No   

  BsmtFinType1  BsmtFinSF1 BsmtFinType2  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          Rec       468.0          LwQ       144.0      270.0        882.0   
1          ALQ       923.0          Unf         0.0      406.0       1329.0   
2          GLQ       791.0          Unf         0.0      137.0        928.0   
3          GLQ       602.0          Unf         0.0      324.0        926.0   
4          ALQ       263.0          Unf         0.0     1017.0       1280.0   

  Heating HeatingQC CentralAir Electrical  1stFlrSF  2ndFlrSF  LowQualFinSF  \
0    GasA        TA          Y      SBrkr       896         0             0   
1    GasA        TA          Y      SBrkr      1329         0             0   
2    GasA        Gd          Y      SBrkr       928       701             0   
3    GasA        Ex          Y      SBrkr       926       678             0   
4    GasA        Ex          Y      SBrkr      1280         0             0   

   GrLivArea  BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  \
0        896           0.0           0.0         1         0             2   
1       1329           0.0           0.0         1         1             3   
2       1629           0.0           0.0         2         1             3   
3       1604           0.0           0.0         2         1             3   
4       1280           0.0           0.0         2         0             2   

   KitchenAbvGr KitchenQual  TotRmsAbvGrd Functional  Fireplaces FireplaceQu  \
0             1          TA             5        Typ           0         NaN   
1             1          Gd             6        Typ           0         NaN   
2             1          TA             6        Typ           1          TA   
3             1          Gd  

In [9]:
df_merge.shape

(2919, 82)

In [10]:
#Get a list of all the categorical features that have the keyword 'Qual' OR 'Cond' OR 'Qu' OR 'QC' in the feature name
feature_rating_Qual = [col for col in df_merge.columns if 'Qual' in col and df_merge[col].dtypes=='object']
feature_rating_Cond = [col for col in df_merge.columns if 'Cond' in col and col not in ['Condition1', 'Condition2', 'SaleCondition'] and df_merge[col].dtypes=='object']
feature_rating_Qu = [col for col in df_merge.columns if 'Qu' in col and df_merge[col].dtypes=='object' and col not in feature_rating_Qual]
feature_rating_QC = [col for col in df_merge.columns if 'QC' in col and df_merge[col].dtypes=='object']

cat_feature_with_rating = feature_rating_Qual + feature_rating_Cond + feature_rating_Qu + feature_rating_QC

for x in cat_feature_with_rating:
    print(x)

ExterQual
BsmtQual
KitchenQual
GarageQual
ExterCond
BsmtCond
GarageCond
FireplaceQu
HeatingQC
PoolQC


In [11]:
#On checking the data description, Missing is valid in some categorical
#Handling Missing Values in Categorical features that have a RATING by replacing them with 'Missing' keyword
df_merge[cat_feature_with_rating] = df_merge[cat_feature_with_rating].fillna("Missing") 

#Making sure the missing values have been handled
print(get_cols_with_missing_values(df_merge[cat_feature_with_rating]))

Series([], dtype: int64)


In [12]:
#Get categorical and numerical columns
categorical_cols=[cname for cname in df_merge.columns if df_merge[cname].dtypes=='object' and cname!='ind']
remaining_cat_cols = [cname for cname in categorical_cols if cname not in cat_feature_with_rating]

numerical_cols=[cname for cname in df_merge.columns if df_merge[cname].dtypes!='object' and cname!='SalePrice']

In [13]:
#Handling Missing Values in Categorical features by replacing them with the feature mode value
for col in remaining_cat_cols:
    df_merge[col] = df_merge[col].fillna(df_merge[col].mode()[0])  


#Making sure the missing values have been handled
print(get_cols_with_missing_values(df_merge[remaining_cat_cols]))

Series([], dtype: int64)


In [14]:
#Handling Missing Values in Numerical features by replacing them with Median value
df_merge[numerical_cols]=df_merge[numerical_cols].fillna(df_merge[numerical_cols].median())

#Making sure the missing values have been handled
print(get_cols_with_missing_values(df_merge[numerical_cols]))

Series([], dtype: int64)


In [15]:
#Select categorical columns with low cardinality
categorical_cols=[cname for cname in df_merge.columns if df_merge[cname].dtypes=='object' and df_merge[cname].nunique()<10]
numerical_cols=[cname for cname in df_merge.columns if df_merge[cname].dtypes!='object']

# Keep selected columns only
my_cols = numerical_cols + categorical_cols
df_merge_clean = df_merge[my_cols].copy()

In [16]:
print(get_cols_with_missing_values(df_merge_clean))

SalePrice    1459
dtype: int64


In [17]:
df_merge_clean.head()

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  \
0  1461          20         80.0    11622            5            6   
1  1462          20         81.0    14267            6            6   
2  1463          60         74.0    13830            5            5   
3  1464          60         78.0     9978            6            6   
4  1465         120         43.0     5005            8            5   

   YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  \
0       1961          1961         0.0       468.0       144.0      270.0   
1       1958          1958       108.0       923.0         0.0      406.0   
2       1997          1998         0.0       791.0         0.0      137.0   
3       1998          1998        20.0       602.0         0.0      324.0   
4       1992          1992         0.0       263.0         0.0     1017.0   

   TotalBsmtSF  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  \
0        882.0       896         0             0        896           0.0   
1       1329.0      1329         0             0       1329           0.0   
2        928.0       928       701             0       1629           0.0   
3        926.0       926       678             0       1604           0.0   
4       1280.0      1280         0             0       1280           0.0   

   BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  \
0           0.0         1         0             2             1             5   
1           0.0         1         1             3             1             6   
2           0.0         2         1             3             1             6   
3           0.0         2         1             3             1             7   
4           0.0         2         0             2             1             5   

   Fireplaces  GarageYrBlt  GarageCars  GarageArea  WoodDeckSF  OpenPorchSF  \
0           0       1961.0         1.0       730.0         140            0   
1           0       1958.0         1.0       312.0         393           36   
2           1       1997.0         2.0       482.0         212           34   
3           1       1998.0         2.0       470.0         360           36   
4           0       1992.0         2.0       506.0           0           82   

   EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  MiscVal  MoSold  YrSold  \
0              0          0          120         0        0       6    2010   
1              0          0            0         0    12500       6    2010   
2              0          0            0         0        0       3    2010   
3              0          0            0         0        0       6    2010   
4              0          0          144         0        0       1    2010   

   SalePrice MSZoning Street Alley LotShape LandContour Utilities LotConfig  \
0        NaN       RH   Pave  Grvl      Reg         Lvl    AllPub    Inside   
1        NaN       RL   Pave  Grvl      IR1         Lvl    AllPub    Corner   
2        NaN       RL   Pave  Grvl      IR1         Lvl    AllPub    Inside   
3        NaN       RL   Pave  Grvl      IR1         Lvl    AllPub    Inside   
4        NaN       RL   Pave  Grvl      IR1         HLS    AllPub    Inside   

  LandSlope Condition1 Condition2 BldgType HouseStyle RoofStyle RoofMatl  \
0       Gtl      Feedr       Norm     1Fam     1Story     Gable  CompShg   
1       Gtl       Norm       Norm     1Fam     1Story       Hip  CompShg   
2       Gtl       Norm       Norm     1Fam     2Story     Gable  CompShg   
3       Gtl       Norm       Norm     1Fam     2Story     Gable  CompShg   
4       Gtl       Norm       Norm   TwnhsE     1Story     Gable  CompShg   

  MasVnrType ExterQual ExterCond Foundation BsmtQual BsmtCond BsmtExposure  \
0       None        TA        TA     CBlock       TA       TA           No   
1    BrkFace        TA        TA     CBlock       TA       TA           No   
2       None        TA        TA      PConc      

# Feature Engineering

In [18]:
#Drop 'Id' feature
df_merge_clean.drop('Id', axis=1, inplace=True)

In data description, the values given for 'OverallQual' feature are as follows- 10:Very Excellent ,9:Excellent ,8:Very Good ,7:Good ,6:Above Average ,5:Average ,4:Below Average ,3:Fair ,2:Poor ,1:Very Poor

So we will try to convert categorical features having qualities as per the above example

In [19]:
#Before encoding - ordinal features
df_merge_clean[cat_feature_with_rating]

ExterQual BsmtQual KitchenQual GarageQual ExterCond BsmtCond GarageCond  \
0           TA       TA          TA         TA        TA       TA         TA   
1           TA       TA          Gd         TA        TA       TA         TA   
2           TA       Gd          TA         TA        TA       TA         TA   
3           TA       TA          Gd         TA        TA       TA         TA   
4           Gd       Gd          Gd         TA        TA       TA         TA   
...        ...      ...         ...        ...       ...      ...        ...   
1455        TA       Gd          TA         TA        TA       TA         TA   
1456        TA       Gd          TA         TA        TA       TA         TA   
1457        Ex       TA          Gd         TA        Gd       Gd         TA   
1458        TA       TA          Gd         TA        TA       TA         TA   
1459        Gd       TA          TA         TA        TA       TA         TA   

     FireplaceQu HeatingQC   PoolQC  
0        Missing        TA  Missing  
1        Missing        TA  Missing  
2             TA        Gd  Missing  
3             Gd        Ex  Missing  
4        Missing        Ex  Missing  
...          ...       ...      ...  
1455          TA        Ex  Missing  
1456          TA        TA  Missing  
1457          Gd        Ex  Missing  
1458     Missing        Gd  Missing  
1459     Missing        Gd  Missing  

[2919 rows x 10 columns]

In [20]:
for col in cat_feature_with_rating:
    if 'Missing' in df_merge_clean[col].value_counts().index:
        df_merge_clean[col] = df_merge_clean[col].map({"Missing":0,"Po":2,"Fa":3,"TA":5,"Gd":7,"Ex":9})
    else:
        df_merge_clean[col] = df_merge_clean[col].map({"Po":2,"Fa":3,"TA":5,"Gd":7,"Ex":9})

In [21]:
#after encoding - ordinal features
df_merge_clean[cat_feature_with_rating]

ExterQual  BsmtQual  KitchenQual  GarageQual  ExterCond  BsmtCond  \
0             5         5            5           5          5         5   
1             5         5            7           5          5         5   
2             5         7            5           5          5         5   
3             5         5            7           5          5         5   
4             7         7            7           5          5         5   
...         ...       ...          ...         ...        ...       ...   
1455          5         7            5           5          5         5   
1456          5         7            5           5          5         5   
1457          9         5            7           5          7         7   
1458          5         5            7           5          5         5   
1459          7         5            5           5          5         5   

      GarageCond  FireplaceQu  HeatingQC  PoolQC  
0              5            0          5       0  
1              5            0          5       0  
2              5            5          7       0  
3              5            7          9       0  
4              5            0          9       0  
...          ...          ...        ...     ...  
1455           5            5          9       0  
1456           5            5          5       0  
1457           5            7          9       0  
1458           5            0          7       0  
1459           5            0          7       0  

[2919 rows x 10 columns]

In [22]:
cat_remaining_to_encode = []

for col in categorical_cols:
    if col not in cat_feature_with_rating and col !='ind':
        cat_remaining_to_encode.append(col)


print(cat_remaining_to_encode)

['MSZoning', 'Street', 'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig', 'LandSlope', 'Condition1', 'Condition2', 'BldgType', 'HouseStyle', 'RoofStyle', 'RoofMatl', 'MasVnrType', 'Foundation', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinType2', 'Heating', 'CentralAir', 'Electrical', 'Functional', 'GarageType', 'GarageFinish', 'PavedDrive', 'Fence', 'MiscFeature', 'SaleType', 'SaleCondition']


In [23]:
#Before encoding
df_merge_clean[cat_remaining_to_encode]

MSZoning Street Alley LotShape LandContour Utilities LotConfig LandSlope  \
0          RH   Pave  Grvl      Reg         Lvl    AllPub    Inside       Gtl   
1          RL   Pave  Grvl      IR1         Lvl    AllPub    Corner       Gtl   
2          RL   Pave  Grvl      IR1         Lvl    AllPub    Inside       Gtl   
3          RL   Pave  Grvl      IR1         Lvl    AllPub    Inside       Gtl   
4          RL   Pave  Grvl      IR1         HLS    AllPub    Inside       Gtl   
...       ...    ...   ...      ...         ...       ...       ...       ...   
1455       RL   Pave  Grvl      Reg         Lvl    AllPub    Inside       Gtl   
1456       RL   Pave  Grvl      Reg         Lvl    AllPub    Inside       Gtl   
1457       RL   Pave  Grvl      Reg         Lvl    AllPub    Inside       Gtl   
1458       RL   Pave  Grvl      Reg         Lvl    AllPub    Inside       Gtl   
1459       RL   Pave  Grvl      Reg         Lvl    AllPub    Inside       Gtl   

     Condition1 Condition2 BldgType HouseStyle RoofStyle RoofMatl MasVnrType  \
0         Feedr       Norm     1Fam     1Story     Gable  CompShg       None   
1          Norm       Norm     1Fam     1Story       Hip  CompShg    BrkFace   
2          Norm       Norm     1Fam     2Story     Gable  CompShg       None   
3          Norm       Norm     1Fam     2Story     Gable  CompShg    BrkFace   
4          Norm       Norm   TwnhsE     1Story     Gable  CompShg       None   
...         ...        ...      ...        ...       ...      ...        ...   
1455       Norm       Norm     1Fam     2Story     Gable  CompShg       None   
1456       Norm       Norm     1Fam     1Story     Gable  CompShg      Stone   
1457       Norm       Norm     1Fam     2Story     Gable  CompShg       None   
1458       Norm       Norm     1Fam     1Story       Hip  CompShg       None   
1459       Norm       Norm     1Fam     1Story     Gable  CompShg       None   

     Foundation BsmtExposure BsmtFinType1 BsmtFinType2 Heating CentralAir  \
0        CBlock           No          Rec          LwQ    GasA          Y   
1        CBlock           No          ALQ          Unf    GasA          Y   
2         PConc           No          GLQ          Unf    GasA          Y   
3         PConc           No          GLQ          Unf    GasA          Y   
4         PConc           No          ALQ          Unf    GasA          Y   
...         ...          ...          ...          ...     ...        ...   
1455      PConc           No          Unf          Unf    GasA          Y   
1456     CBlock           No          ALQ          Rec    GasA          Y   
1457      Stone           No          GLQ          Unf    GasA          Y   
1458     CBlock           Mn          GLQ          Rec    GasA          Y   
1459     CBlock           No          BLQ          LwQ    GasA          Y   

     Electrical Functional GarageType GarageFinish PavedDrive  Fence  \
0         SBrkr        Typ     Attchd          Unf          Y  MnPrv   
1         SBrkr        Typ     Attchd          Unf          Y  MnPrv   
2         SBrkr        Typ     Attchd          Fin          Y  MnPrv   
3         SBrkr        Typ     Attchd          Fin          Y  MnPrv   
4         SBrkr        Typ     Attchd          RFn          Y  MnPrv   
...         ...        ...        ...          ...        ...    ...   
1455      SBrkr        Typ     Attchd          RFn          Y  MnPrv   
1456      SBrkr       Min1     Attchd          Unf          Y  MnPrv   
1457      SBrkr        Typ     Attchd          RFn          Y  GdPrv   
1458      FuseA        Typ     Attchd          Unf          Y  MnPrv   
1459      SBrkr        Typ     Attchd          Fin          Y  MnPrv   

     MiscFeature SaleType SaleCondition  
0           Shed       WD        Normal  
1           Gar2       WD        Normal  
2           Shed       WD        Normal  
3           Shed       WD        Normal  
4           Shed       WD        Normal  
...          ...      ...    

In [24]:
df_merge_clean_dummies = pd.get_dummies(df_merge_clean[cat_remaining_to_encode],drop_first=True)

df_merge_clean.drop(cat_remaining_to_encode,axis=1,inplace=True)

df_merge_clean = pd.concat([df_merge_clean,df_merge_clean_dummies],axis=1)

In [25]:
df_merge_clean.head()

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0          20         80.0    11622            5            6       1961   
1          20         81.0    14267            6            6       1958   
2          60         74.0    13830            5            5       1997   
3          60         78.0     9978            6            6       1998   
4         120         43.0     5005            8            5       1992   

   YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  TotalBsmtSF  \
0          1961         0.0       468.0       144.0      270.0        882.0   
1          1958       108.0       923.0         0.0      406.0       1329.0   
2          1998         0.0       791.0         0.0      137.0        928.0   
3          1998        20.0       602.0         0.0      324.0        926.0   
4          1992         0.0       263.0         0.0     1017.0       1280.0   

   1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  BsmtHalfBath  \
0       896         0             0        896           0.0           0.0   
1      1329         0             0       1329           0.0           0.0   
2       928       701             0       1629           0.0           0.0   
3       926       678             0       1604           0.0           0.0   
4      1280         0             0       1280           0.0           0.0   

   FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  TotRmsAbvGrd  Fireplaces  \
0         1         0             2             1             5           0   
1         1         1             3             1             6           0   
2         2         1             3             1             6           1   
3         2         1             3             1             7           1   
4         2         0             2             1             5           0   

   GarageYrBlt  GarageCars  GarageArea  WoodDeckSF  OpenPorchSF  \
0       1961.0         1.0       730.0         140            0   
1       1958.0         1.0       312.0         393           36   
2       1997.0         2.0       482.0         212           34   
3       1998.0         2.0       470.0         360           36   
4       1992.0         2.0       506.0           0           82   

   EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  MiscVal  MoSold  YrSold  \
0              0          0          120         0        0       6    2010   
1              0          0            0         0    12500       6    2010   
2              0          0            0         0        0       3    2010   
3              0          0            0         0        0       6    2010   
4              0          0          144         0        0       1    2010   

   SalePrice  ExterQual  ExterCond  BsmtQual  BsmtCond  HeatingQC  \
0        NaN          5          5         5         5          5   
1        NaN          5          5         5         5          5   
2        NaN          5          5         7         5          7   
3        NaN          5          5         5         5          9   
4        NaN          7          5         7         5          9   

   KitchenQual  FireplaceQu  GarageQual  GarageCond  PoolQC   ind  \
0            5            0           5           5       0  test   
1            7            0           5           5       0  test   
2            5            5           5           5       0  test   
3            7            7           5           5       0  test   
4            7            0           5           5       0  test   

   MSZoning_FV  MSZoning_RH  MSZoning_RL  MSZoning_RM  Street_Pave  \
0            0            1            0            0            1   
1            0            0            1            0            1   
2            0            0            1            0            1   
3            0            0            1            0            1   
4            0            0            1            0            1   

In [26]:
df_merge_clean['GarageYrBlt']=df_merge_clean['GarageYrBlt'].astype('int')

In [27]:
df_merge_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2919 entries, 0 to 1459
Columns: 170 entries, MSSubClass to SaleCondition_Partial
dtypes: float64(11), int64(36), object(1), uint8(122)
memory usage: 1.4+ MB


# Split Train and Test data

In [28]:
test, train= df_merge_clean[df_merge_clean["ind"].eq("test")], df_merge_clean[df_merge_clean["ind"].eq("train")]
test.drop(["SalePrice", "ind"], axis=1, inplace=True)
train.drop(["ind"], axis=1, inplace=True)

/opt/conda/lib/python3.7/site-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [29]:
train.shape, test.shape

((1460, 169), (1459, 168))

In [30]:
X = train.loc[:, train.columns!='SalePrice']
y = train['SalePrice']

# Break off validation set from training data
X_train, X_valid, y_train, y_valid = train_test_split(X, y, train_size=0.7, test_size=0.3, random_state=0)

# summarize
print('Train', X_train.shape, y_train.shape)
print('Validation', X_valid.shape, y_valid.shape)

Train (1021, 168) (1021,)
Validation (438, 168) (438,)


# Feature Selection

Check for features with zero variance

In [31]:
from sklearn.feature_selection import VarianceThreshold as VT
var_thres = VT(threshold=0)
var_thres.fit(X_train)

#get features that have constant values
const_cols = [ col for col in X_train.columns if col not in X_train.columns[var_thres.get_support()]]
print(const_cols)

['Condition2_PosA', 'Condition2_RRAn', 'Condition2_RRNn', 'RoofMatl_Membran', 'Functional_Sev', 'MiscFeature_TenC']


In [32]:
X_train.drop(columns=const_cols, axis=1, inplace=True)
X_valid.drop(columns=const_cols, axis=1, inplace=True)
test.drop(columns=const_cols, axis=1, inplace=True)

/opt/conda/lib/python3.7/site-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [33]:
def correlation(dataset, threshold):
    col_corr = set()
    corr_matrix = dataset.corr()
    for i in range(len(corr_matrix.columns)):
        for j in range(i):
            if abs(corr_matrix.iloc[i,j]) > threshold:
                colname = corr_matrix.columns[i]
                col_corr.add(colname)
    return col_corr

corr_features = correlation(X_train, 0.85)
print("Features with high correlation ", corr_features)

Features with high correlation  {'PoolQC', 'GarageCond', 'SaleCondition_Partial', 'GarageArea', 'RoofStyle_Hip', 'FireplaceQu'}


In [34]:
#Drop features with high correlation
X_train.drop(columns=corr_features, axis=1, inplace=True)
X_valid.drop(columns=corr_features, axis=1, inplace=True)
test.drop(columns=corr_features, axis=1, inplace=True)

In [35]:
X_train.head()

MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
64            60         68.0     9375            7            5       1997   
682          120         68.0     2887            6            5       1996   
960           20         50.0     7207            5            7       1958   
1384          50         60.0     9060            6            5       1939   
1100          30         60.0     8400            2            5       1920   

      YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  BsmtUnfSF  \
64            1998       573.0       739.0         0.0      318.0   
682           1997         0.0      1003.0         0.0      288.0   
960           2008         0.0       696.0         0.0      162.0   
1384          1950         0.0       204.0         0.0      356.0   
1100          1950         0.0       290.0         0.0        0.0   

      TotalBsmtSF  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  BsmtFullBath  \
64         1057.0      1057       977             0       2034           1.0   
682        1291.0      1291         0             0       1291           1.0   
960         858.0       858         0             0        858           1.0   
1384        560.0       698       560             0       1258           0.0   
1100        290.0       438         0             0        438           0.0   

      BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  \
64             0.0         2         1             3             1   
682            0.0         1         0             2             1   
960            0.0         1         0             2             1   
1384           0.0         1         0             2             1   
1100           0.0         1         0             1             1   

      TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  WoodDeckSF  \
64               8           0         1998         2.0         576   
682              6           1         1996         2.0         307   
960              4           0         1979         0.0         117   
1384             6           0         1939         1.0           0   
1100             3           0         1930         1.0           0   

      OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  MiscVal  \
64             36              0          0            0         0        0   
682             0              0          0            0         0        0   
960             0              0          0            0         0        0   
1384            0              0          0            0         0        0   
1100            0              0          0            0         0        0   

      MoSold  YrSold  ExterQual  ExterCond  BsmtQual  BsmtCond  HeatingQC  \
64         2    2009          5          5         7         5          9   
682       11    2008          5          5         7         5          9   
960        2    2010          5          7         5         5          7   
1384      10    2009          5          5         5         5          5   
1100       1    2009          5          5         5         3          5   

      KitchenQual  GarageQual  MSZoning_FV  MSZoning_RH  MSZoning_RL  \
64              7           5            0            0            1   
682             7           5            0            0            1   
960             5           0            0            0            1   
1384            5           5            0            0            1   
1100            3           5            0            0            1   

      MSZoning_RM  Street_Pave  Alley_Pave  LotShape_IR2  LotShape_IR3  \
64              0            1           0             0             0   
682             0            1           0             0             0   
960             0            1           0             0             0   
1384            0            1           0             0             0   
1100            0            1           0    

# Build Model

In [36]:
from xgboost import XGBRegressor
from sklearn.model_selection import RandomizedSearchCV

# A parameter grid for XGBoost
params = {
    'n_estimators':[100,200,400,600,800,900,1000],
    'learning_rate':[0.05,0.1,0.15],
    'min_child_weight':[2*i+1 for i in range(0,5)], 
    'gamma':[0.1*i for i in range(1,10)],  
    'subsample':[0.1*i for i in range(6,11)],
    'colsample_bytree':[0.1*i for i in range(6,11)], 
    'max_depth': [2,3,4,6,7],
    'objective': ['reg:squarederror', 'reg:tweedie'],
    'booster': ['gbtree', 'gblinear'],
    'eval_metric': ['rmse'],
}

regressor = XGBRegressor()
#RandomizedSearchCV
random_search = RandomizedSearchCV(regressor, param_distributions=params,n_iter=5, n_jobs=-1, cv=5, scoring='neg_mean_squared_error')

random_search.fit(X_train, y_train)

RandomizedSearchCV(cv=5,
                   estimator=XGBRegressor(base_score=None, booster=None,
                                          colsample_bylevel=None,
                                          colsample_bynode=None,
                                          colsample_bytree=None,
                                          enable_categorical=False, gamma=None,
                                          gpu_id=None, importance_type=None,
                                          interaction_constraints=None,
                                          learning_rate=None,
                                          max_delta_step=None, max_depth=None,
                                          min_child_weight=None, missing=nan,
                                          monotone_constraints=...
                                        'gamma': [0.1, 0.2, 0.30000000000000004,
                                                  0.4, 0.5, 0.6000000000000001,
                              

In [37]:
best_regressor = random_search.best_estimator_
print(best_regressor)

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=0.9, enable_categorical=False,
             eval_metric='rmse', gamma=0.4, gpu_id=-1, importance_type=None,
             interaction_constraints='', learning_rate=0.1, max_delta_step=0,
             max_depth=4, min_child_weight=1, missing=nan,
             monotone_constraints='()', n_estimators=1000, n_jobs=4,
             num_parallel_tree=1, predictor='auto', random_state=0, reg_alpha=0,
             reg_lambda=1, scale_pos_weight=1, subsample=0.7000000000000001,
             tree_method='exact', validate_parameters=1, verbosity=None)


In [38]:
from sklearn.metrics import mean_absolute_error
# Get predictions
y_pred = best_regressor.predict(X_valid)

In [39]:
# Calculate MAE
rmse_pred = mean_absolute_error(y_valid, y_pred) 
print("Root Mean Absolute Error:" , np.sqrt(rmse_pred))

Root Mean Absolute Error: 128.4682371368645


In [40]:
# Get test predictions
preds_test = best_regressor.predict(test)

In [41]:
# Save test predictions to file
output = pd.DataFrame({'Id': test.index+1461,
                       'SalePrice': preds_test})
output.to_csv('submission.csv', index=False)